File: web/src/api/competition.ts
import type {
  CompetitionListPayload,
  EquityHistoryBatchPayload,
  EquityHistoryBatchRequest,
  EquityHistoryPointPayload,
  EquityHistoryQuery,
  PublicCompetitionTraderPayload,
  PublicTraderConfigPayload,
} from "@/types/public"
import request from "@/utils/request"

const Api = {
  Competition: "/api/competition",
  TopTraders: "/api/competition/top-traders",
  EquityHistory: "/api/competition/equity-history",
  EquityHistoryBatch: "/api/competition/equity-history-batch",
  PublicTraderConfig: "/api/competition/traders/{id}/public-config",
} as const

export function getCompetitionApi() {
  return request.get<CompetitionListPayload>(Api.Competition)
}

export function getTopTradersApi() {
  return request.get<PublicCompetitionTraderPayload[]>(Api.TopTraders)
}

export function getEquityHistoryApi(params?: EquityHistoryQuery) {
  return request.get<EquityHistoryPointPayload[]>(Api.EquityHistory, { params })
}

export function getEquityHistoryBatchApi(data: EquityHistoryBatchRequest) {
  return request.post<EquityHistoryBatchPayload>(Api.EquityHistoryBatch, data)
}

export function getPublicTraderConfigApi(id: string) {
  return request.get<PublicTraderConfigPayload>(
    Api.PublicTraderConfig.replace("{id}", id),
  )
}

File: core/src/http/handlers/competition.rs
use axum::{
    Json,
    extract::{Path, Query, State},
};

use crate::{
    contracts::public::{
        CompetitionListPayload, EquityHistoryBatchPayload, EquityHistoryBatchRequest,
        EquityHistoryPointPayload, EquityHistoryQuery, PublicCompetitionTraderPayload,
        PublicTraderConfigPayload,
    },
    error::Result,
    http::response::ApiResponse,
    state,
};

pub async fn handle_public_competition(
    State(app): State<state::AppState>,
) -> Result<Json<ApiResponse<CompetitionListPayload>>> {
    let traders = app.services.competition_service.competition().await?;
    let payload = CompetitionListPayload {
        count: traders.len(),
        traders,
    };
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_top_traders(
    State(app): State<state::AppState>,
) -> Result<Json<ApiResponse<Vec<PublicCompetitionTraderPayload>>>> {
    let payload = app.services.competition_service.top_traders().await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_equity_history(
    State(app): State<state::AppState>,
    Query(q): Query<EquityHistoryQuery>,
) -> Result<Json<ApiResponse<Vec<EquityHistoryPointPayload>>>> {
    let payload = app.services.competition_service.equity_history(q).await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_equity_history_batch(
    State(app): State<state::AppState>,
    Json(request): Json<EquityHistoryBatchRequest>,
) -> Result<Json<ApiResponse<EquityHistoryBatchPayload>>> {
    let payload = app
        .services
        .competition_service
        .equity_history_batch(request)
        .await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_public_trader_config(
    State(app): State<state::AppState>,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<PublicTraderConfigPayload>>> {
    let payload = app
        .services
        .competition_service
        .public_trader_config(&id)
        .await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}
